File: 2015/swift/4/main.swift
import Foundation
import CryptoKit

func main(){
    let args = CommandLine.arguments

    if args.count < 2 {
        print("Usage: swift main.swift <input.txt>")
        return
    }

    let filename = args[1]
    let file = FileHandle(forReadingAtPath: filename)

    if file == nil {
        print("Cannot open file")
        return
    }

    let data = file!.readDataToEndOfFile()
    let prefix_text = String(data: data, encoding: .utf8)!

    var i = 0
    var leading_5_found = false

    while(true){
        let input = prefix_text + String(i)
        let hash = Insecure.MD5.hash(data: Data(input.utf8))

        let hash_string = hash.map { String(format: "%02hhx", $0) }.joined()

        if (hash_string.starts(with: "000000") ){
            print("Part 2: \(i)")
            
            break
        }

        if(hash_string.starts(with: "00000") && !leading_5_found){
            print("Part 1: \(i)")
            leading_5_found = true
        }

        i+=1
    }

}

main()

File: 2015/php/4/main.php
<?php

function main(){
    $args = $_SERVER['argv'];
    if (count($args) < 2){
        echo "Usage: php main.php <input.txt>\n";
        return;
    }

    $filename = $args[1];
    $file = fopen($filename, "r");
    if ($file == false){
        echo "Error: Cannot open file $filename\n";
        return;
    }

    $line = fgets($file);

    $i = 0;
    $leading_5_found = false;
    while(true){
        $input = $line . $i;
        $hash = md5($input);

        if (substr($hash, 0, 6) === "000000"){
            echo "Part 2: $i\n";
            break;
        }

        if (substr($hash, 0, 5) === "00000" && !$leading_5_found){
            echo "Part 1: $i\n";
            $leading_5_found = true;
        }

        $i++;
    }
}

main();